[Oracle Module] New sysmetric metricset

yug-elastic opened the pull request:
- Enhancement

## What does this PR do?

Create a new sysmetric metricset to complement the current performance one.

## Why is it important?

What Oracle DBAs usually do is actually pull the pre-calculated metrics from the V$SYSMETRIC table (called System View). DBAs usually need around 10 out of these 200+ metrics and Oracle will keep adding more metrics to this table in the future, so, there has to be a way to dynamically filter in or query only those that are suggested by the user with the help of a keyword or Regex.

## Checklist

- [x] My code follows the style guidelines of this projects
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #30946

## Screenshot
![Sysmetric_Dashboard](https://user-images.githubusercontent.com/96041884/166219074-1738a1f6-d16a-4d6b-9932-353bd10c6dc6.png)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31462/detail/PR-31462/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31462/detail/PR-31462/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31462/detail/PR-31462/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31462/detail/PR-31462/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_31462.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-04-28T18:38:41.020Z&rangeTo=2022-04-28T18:58:41.020Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=2bde8dd1c78e99085a9dbb55ee9f0224&transactionId=529ec9e8e2544407)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-04-28T18:48:41.020+0000


* Duration: 14 min 4 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-31462/detail/PR-31462/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `x-pack/metricbeat-checks - make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -`
<ul>
<li>Took 2 min 3 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31462/runs/1/steps/1090/log/?start=0">here</a></li>
<li>Description: <code>make -C x-pack/metricbeat check;make -C x-pack/metricbeat update;make -C metricbeat check;make -C metricbeat update;make check-no-changes;</code></l1>
</ul>
  
  
  
  
##### `make check-default`
<ul>
<li>Took 1 min 29 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31462/runs/1/steps/837/log/?start=0">here</a></li>
<li>Description: <code>make check-default</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-31462/runs/1/steps/1099/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 2'</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

cmacknz commented on x-pack/metricbeat/module/oracle/sysmetric/sysmetric_metric.go:
```diff
@@ -0,0 +1,89 @@
+ *
+ * Which is parsed into sysmetricMetric instances
+ */
+func (e *sysmetricExtractor) calQuery() string {
```
```suggestion
func (e *sysmetricExtractor) callQuery() string {
```

I suspect there is a missing `l` here

yug-elastic replied:
`calQuery` was actually supposed to mean "calculate query". Should we rename it to `calculateQuery` or do you suggest rephrasing it?

cmacknz replied:
Ah yes `calculateQuery` is fine. Thanks. `calc` is a more frequent abbreviation for `calculate` than `cal`.

yug-elastic replied:
That makes sense, we'll make sure to use the same going ahead. Thanks!

cmacknz commented on x-pack/metricbeat/module/oracle/sysmetric/sysmetric_metric.go:
```diff
@@ -0,0 +1,89 @@
+		e.patterns = []string{"%"}
+	}
+
+	query := "SELECT * FROM V$SYSMETRIC WHERE (" + "METRIC_NAME LIKE '" + e.patterns[0] + "'"
```
Do we need to sanitize the `e.patterns` here? Could this be done with a bound parameter instead so that arbitrary SQL from the config file can't just be inserted into this statement?

yug-elastic replied:
Right, updated the PR with this change. Thank you!

cmacknz commented on x-pack/metricbeat/module/oracle/_meta/docs.asciidoc:
```diff
@@ -1,36 +1,36 @@
-=== `tablespaces`
-
-Includes information about data files and temp files, grouped by Tablespace with free space available, used space, status of the data files, status of the Tablespace, etc.
+This is the https://www.oracle.com[Oracle] module for Metricbeat. It is under active development with feedback from the community. A single Metricset for Tablespace monitoring is added so the community can start gathering metrics from their nodes and contributing to the module.
```
Did anything change in the documentation here? I see the whole being replaces with what looks like the same content.

yug-elastic replied:
I think this is because of `LF` → `CLRF`, let me revert it. I'll also update the documentation with respect to `sysmetric` metricset.

cmacknz commented:
You'll need to add a CHANGELOG entry for this.

I'm not very familiar with this code. The commit history (https://github.com/elastic/beats/commits/main/x-pack/metricbeat/module/oracle) suggests @sayden would be a good reviewer here.

cmacknz commented:
This LGTM to me, but I am not very familiar with the Oracle modules. Hopefully @sayden can do a detailed review or would know someone who can.

yug-elastic commented:
> This LGTM to me, but I am not very familiar with the Oracle modules. Hopefully @sayden can do a detailed review or would know someone who can.

Thanks for the approval and the clarification, @cmacknz!

@sayden, Please help us with the review or let me know if I should request review from someone from the team who would be acquainted with Oracle module.

yug-elastic commented:
Thanks for taking out time to review this PR, @sayden!

> None of the fields in the PR follow Elastic naming conventions: https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html and the timestamps must be converted to ISO8601 if you don´t want to provide a `format` in the date type.

As a part of this enhancement, we had to map the fields dynamically as mentioned in the [PRD](https://github.com/elastic/beats/issues/30946). Your point makes sense, we can add a function to transform the field names dynamically to snake case to follow the Elastic naming conventions and convert timestamps to date. But because of the dynamic mapping, it does not seem feasible to group the fields.

> DBAs usually need around 10 out of these 200+ metrics and Oracle will keep adding more metrics to this table in the future, so, there has to be a way to dynamically filter in or query only those that are suggested by the user with the help of a keyword or Regex.

yug-elastic commented:
Hey @sayden, we have addressed your review comments and made some changes. We need your inputs on the other comments, can you please re-review the PR?

matschaffer commented:
Looks like @elastic/infra-monitoring-ui was added by mistake so I've removed it from the reviewers list.

kush-elastic reviewed (approved):
There are few nits from my side and there are few methods that can be explained a bit using comments. So others can easily understood the workflow.
Else LGTM!

agithomas commented on x-pack/metricbeat/module/oracle/sysmetric/data.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+// extract is the E of a ETL processing. Gets all the system metric information into an instance of extractedData
+func (m *MetricSet) extract(ctx context.Context, extractor sysmetricExtractMethod) (out *extractedData, err error) {
```
From the definition of ETL, it stands for extract, transform and load, is a data integration process that combines data from multiple data sources into a single, consistent data store that is loaded into a  datawarehouse. It is a term commonly used in datawarehousing domain not in observability 

I understand that similar function names is used for performance and tablespace. 

I do not not think that it is a correct name to be followed for an observability product. 



Recommendation : Suggest to use different function names for extract, extranctAndTransform and transform. Also, avoid comments that says that this is a ETL specific operation that is happening.

agithomas commented:
Second part of the requirement is as below. 

Another problem with the queries used by the metricbeat module is that they don’t account for the rack installations. If there is a rack-installation, metricbeat is not going to provide correct results, because nobody knows whether the data reported is from the current node that you connected to or is an average across all nodes in the rack.

How this requirement is met?

yug-elastic commented on metricbeat/docs/modules/oracle.asciidoc:
```diff
@@ -80,22 +86,29 @@ in <<configuration-metricbeat>>. Here is an example configuration:
+  metricsets:
+    - sysmetric
+  enabled: true
+  # patterns: ["foo%","%bar","%foobar%"]
```
We have mentioned the following as a part of the documentation for `patterns` based on our suggestions from @akshay-saraswat during the TDD walkthrough:

This metricset dynamically filters metrics based on given input `patterns`. These patterns can be keywords or regular expressions. The `sysmetric` metricset uses the `OR` operator when forming the SQL query internally. Refer [here](https://docs.oracle.com/cd/B12037_01/server.101/b10759/conditions016.htm) for more details.

Please suggest what more details you think we should add here.

agithomas commented:
> > Second part of the requirement is as below.
> > Another problem with the queries used by the metricbeat module is that they don’t account for the rack installations. If there is a rack-installation, metricbeat is not going to provide correct results, because nobody knows whether the data reported is from the current node that you connected to or is an average across all nodes in the rack.
> > How this requirement is met?
> 
> We have covered all the points mentioned as a part of [solution](https://github.com/elastic/beats/issues/30946#:~:text=in%20the%20rack.-,Solution,-What%20Oracle%20DBAs) in the PRD.
> 
> If we refer to the `V$SYSMETRIC` table documentation [here](https://docs.oracle.com/en/database/oracle/oracle-database/21/refrn/V-SYSMETRIC.html#GUID-623748C3-F765-4149-8378-F5CDAD59909A), there is no field which can help us in identifying the node from which the data is reported. We also verified the metrics and units from the output of sysmetric table, but could not find an identifier. So, it does not seem to be possible to cover that as a part of sysmetric metricset.
> 
> Do you have any ideas over it?

Please confirm if @akshay-saraswat  is informed about proceeding with the reduced scope of not having the RACK details? 

If the scope is not changed, a solution must be found. Please confirm.

agithomas commented:
> > Second part of the requirement is as below.
> > Another problem with the queries used by the metricbeat module is that they don’t account for the rack installations. If there is a rack-installation, metricbeat is not going to provide correct results, because nobody knows whether the data reported is from the current node that you connected to or is an average across all nodes in the rack.
> > How this requirement is met?
> 
> We have covered all the points mentioned as a part of [solution](https://github.com/elastic/beats/issues/30946#:~:text=in%20the%20rack.-,Solution,-What%20Oracle%20DBAs) in the PRD.
> 
> If we refer to the `V$SYSMETRIC` table documentation [here](https://docs.oracle.com/en/database/oracle/oracle-database/21/refrn/V-SYSMETRIC.html#GUID-623748C3-F765-4149-8378-F5CDAD59909A), there is no field which can help us in identifying the node from which the data is reported. We also verified the metrics and units from the output of sysmetric table, but could not find an identifier. So, it does not seem to be possible to cover that as a part of sysmetric metricset.
> 
> Do you have any ideas over it?

You may check V$INSTANCE especially INSTANCE_MODE value. It might help.

yug-elastic commented:
Hey @agithomas, we have made all the changes as per your suggestions and discussion over the sync. We are looking into the parameter CON_ID by doing some hands-on with the Oracle Database. We will get back to you with our analysis over the same soon and we can agree upon something after discussing with the team.

Please feel free to review the rest of the PR and let us know if you have any other suggestions/feedback.

agithomas reviewed (approved):
LGTM
